fendermint/vm/interpreter/src/fvm/query.rs

```diff
@@ -99,3 +96,131 @@ where
+                "message execution failed with error code {}",
+                ret.msg_receipt.exit_code
+            ));
+        }
+        out.gas_limit = (ret.msg_receipt.gas_used as f64 * Self::GAS_OVERESTIMATION_RATE) as u64;
+        if out.gas_premium.is_zero() {
+            // TODO: Instead of assigning a default value here, we should analyze historical
+            // blocks from the current height to estimate an accurate value for this premium.
+            // To achieve this we would need to perform a set of ABCI queries.
+            // In the meantime, this value should be good enough to make sure that the
+            // message is included in a block.
+            out.gas_premium = TokenAmount::from_nano(BigInt::from(Self::DEFAULT_GAS_PREMIUM));
+        }
+        if out.gas_fee_cap.is_zero() {
+            // Compute the fee cap from gas premium and applying an additional overestimation.
+            let overestimated_limit = (out.gas_limit as f64 * Self::GAS_OVERESTIMATION_RATE) as u64;
+            out.gas_fee_cap = std::cmp::min(
+                TokenAmount::from_atto(BigInt::from(overestimated_limit)) + &out.gas_premium,
+                TokenAmount::from_atto(BLOCK_GAS_LIMIT),
+            );
+
+            // TODO: In Lotus historical values of the base fee and a more accurate overestimation is performed
+            // for the fee cap. If we issues with messages going through let's consider the historical analysis.
+        }
+
+        Ok(out)
+    }
+
+    // This function performs a simpler implementation of the gas search than the one used in Lotus.
+    // Instead of using historical information of the gas limit for other messages, it searches
+    // for a valid gas limit for the current message in isolation.
+    fn gas_search(
+        &self,
+        state: &FvmQueryState<DB>,
+        msg: &mut Message,
+    ) -> anyhow::Result<GasEstimate> {
+        let mut curr_limit = msg.gas_limit;
+
+        while {
+            if let Some(ret) = self.estimation_call_with_limit(state, msg, curr_limit)? {
+                return Ok(ret);
+            }
+
+            curr_limit = (curr_limit as f64 * Self::GAS_SEARCH_STEP) as u64;
+            if curr_limit > BLOCK_GAS_LIMIT {
+                return Ok(GasEstimate {
+                    exit_code: ExitCode::OK,
+                    info: "".to_string(),
+                    gas_limit: BLOCK_GAS_LIMIT,
+                });
+            }
+            true
+        } {}
+
+        // TODO: For a more accurate gas estimation we could track the low and the high
+        // of the search and make higher steps (e.g. `GAS_SEARCH_STEP = 2`).
+        // Once an interval is found of [low, high] for which the message
+        // succeeds, we make a finer-grained within that interval.
+        // At this point, I don't think is worth being that accurate as long as it works.
+
+        Err(anyhow::anyhow!(
+            "gas search failed. no valid gas limit found"
+        ))
```

Comment:
This will never be called because you never break out of the loop, only return.

Reply:
Fixed.